File: matlab/classification/Identification_procedure/identification_general.py
import numpy as np
import time

def identification_general(database, aux_L, prior_g, my_class, param):
    """
    Translated from identification_general.m
    """
    N_spect = param['N_spect']
    gpuFlag = param.get('gpuFlag', 0)
    n_buff = param.get('n_buff', 1)
    is_dict = param.get('dict', 1)
    
    N = database.shape[0]
    dict_size = aux_L.shape[0]
    
    database = database.T
    aux_L_comp = aux_L[:, :N_spect].T
    
    n_class = len(my_class)
    model_size = np.zeros(n_class)
    for k in range(n_class):
        model_size[k] = np.sum(aux_L[:, -2] == (k + 1))
        
    cum_model_size = np.cumsum(model_size).astype(int)
    msize = np.concatenate(([0], cum_model_size))
    
    # Avoid log(0)
    prior_g = np.maximum(prior_g, 1e-10)
    prior_g = np.log(prior_g / np.sum(prior_g))
    
    L_bay = np.zeros(N)
    L = np.zeros(dict_size)
    likelihood_group = np.zeros(len(msize) - 1)
    L_prior_max = np.zeros(len(msize) - 1)
    posterior_g = np.zeros((N, n_class))
    computation_time = np.zeros(N)
    
    for b in range(N):
        start_time = time.time()
        
        if is_dict:
            spectrum = database[:-2, b]
        else:
            data = database[:-2, b]
            spectrum = np.abs(np.fft.fft(data)) ** 2
            
        sum_spec = np.sum(spectrum[:N_spect])
        if sum_spec > 0:
            spectrum_norm = N_spect * spectrum[:N_spect] / sum_spec
        else:
            spectrum_norm = spectrum[:N_spect]
            
        # Pointwise multiplication of matching shapes and sum across freq dimension
        rspectrum_norm = np.tile(spectrum_norm, (dict_size, 1)).T
        L = np.sum(rspectrum_norm * aux_L_comp, axis=0) 
        
        for ii in range(len(msize) - 1):
            A = L[msize[ii]:msize[ii+1]]
            if len(A) > 0:
                L_prior_max[ii] = np.max(A)
                likelihood_group[ii] = np.log(np.nansum(np.exp(A - L_prior_max[ii])))
            else:
                L_prior_max[ii] = -np.inf
                likelihood_group[ii] = -np.inf
                
        likelihood_group = likelihood_group + L_prior_max - np.log(np.maximum(model_size, 1e-10))
        
        A = likelihood_group + prior_g
        L_prior_max_g = np.max(A)
        norm_factor_g = L_prior_max_g + np.log(np.sum(np.exp(A - L_prior_max_g)))
        posterior_g[b, :] = -norm_factor_g + likelihood_group + prior_g
        
        computation_time[b] = time.time() - start_time
        
        if (b + 1) % n_buff == 0:
            sum_g = np.nansum(posterior_g[b - n_buff + 1: b + 1, :], axis=0)
            max_idx = np.where(sum_g == np.max(sum_g))[0][0]
            L_bay[b - n_buff + 1: b + 1] = my_class[max_idx]
            
    return L_bay, posterior_g, computation_time

File: matlab/classification/Identification_procedure/identification_general.m
function [L_bay,posterior_g,computation_time] = identification_general(database, aux_L, prior_g, my_class, param)
% Set parameters
N_spect = param.N_spect;
%type = param.type;
gpuFlag = param.gpuFlag;
n_buff = param.n_buff;

N = size(database,1);
dict_size = size(aux_L,1);

database = database';

% Precompute things
if gpuFlag
    aux_L_comp = gpuArray(aux_L(:,1:N_spect)');
else
    aux_L_comp = aux_L(:,1:N_spect)';
end

n_class = length(my_class);
% get model size
model_size = zeros(1,n_class);
for k = 1:n_class
    model_size(k) = sum(aux_L(:,end-1) == k);
end
cum_model_size = cumsum(model_size);

msize = [0, cum_model_size];

prior_g = log(prior_g ./ sum(prior_g));

% Initialize variables
L_bay = zeros(1,N);
if gpuFlag
    L = zeros(1, dict_size,'gpuArray');
    likelihood_group = zeros(1,length(msize)-1,'gpuArray');
    L_prior_max = zeros(1,length(msize)-1,'gpuArray');
    posterior_g = zeros(N, n_class,'gpuArray');
else
    L = zeros(1, dict_size);
    likelihood_group = zeros(1,length(msize)-1);
    L_prior_max = zeros(1,length(msize)-1);
    posterior_g = zeros(N, n_class);
end

% Identification
computation_time = zeros(1,N);
for b = 1:N
    tic
    if param.dict
        spectrum = database(1:end-2,b);
    else
        % Calcul du spectre du buffer
        data = database(1:end-2,b);
        % Compute spectrum
        spectrum = abs(fft(data)).^2;
    end
%     if 20*log10(sum(abs(data).^2)) < -60
%         posterior_g(b,:) = NaN;
%         continue;
%     else
    spectrum_norm = N_spect * spectrum(1:N_spect) ./ sum(spectrum(1:N_spect));%N_spect * spectrum(1:N_spect) ./ sum(spectrum(1:N_spect));
    if gpuFlag
        spectrum_norm = gpuArray(spectrum_norm);
    end
    
    rspectrum_norm = repmat(spectrum_norm, [1,dict_size]);
    L = sum(rspectrum_norm .* aux_L_comp);
    
    for ii = 1:length(msize)-1
        A = L(msize(ii)+1:msize(ii+1));
        L_prior_max(ii) = max(A);
        if gpuFlag
            likelihood_group(ii) = gpuArray(log(sum(gather(exp(A - L_prior_max(ii))),'omitnan')));
        else
            likelihood_group(ii) = log(sum(exp(A - L_prior_max(ii)),'omitnan'));
        end
    end
    likelihood_group = likelihood_group + L_prior_max - log(model_size);
    
    A = likelihood_group + prior_g;
    L_prior_max = max(A);
    norm_factor_g = L_prior_max + log(sum(exp(A - L_prior_max)));
    posterior_g(b,:) = -norm_factor_g + likelihood_group + prior_g;
    %end
    computation_time(b) = toc;
   
    % Cherche la classe correspondance
%     if mod(b, n_buff) == 0
%         if gpuFlag
%             sum_g = sum(gather(posterior_g((b-n_buff+1):b,:)),1,'omitnan');
%         else
%             sum_g = sum(posterior_g((b-n_buff+1):b,:),1,'omitnan');
%         end
%         L_bay((b-n_buff+1):b) = my_class(max(sum_g) == sum_g);
%         clc
%         disp(['iter: ',num2str(b),' / ',num2str(N)])
progressbar(b,N)
%     end
%     
end
